### ISSUE TITLE
Add minimum/maximum functions

### ISSUE THREAD

**Issue by magnus-madsen** [flix/flix#1636]
Add the following functions to the appropriate namespaces e.g. `List`, `MutList`, `Set`, `MutSet`, `Nel`, `Map` (is it by key or value? check Haskell), `Array`

- [ ] `minimum[a: Order](l: List[a]): Option[a]`
- [ ] `maximum[a: Order](l: List[a]): Option[a]`
- [ ] `minimumBy(cmp: (a, a) -> Comparison,l: List[a]): Option[a]`
- [ ] `maximumBy(cmp: (a, a) -> Comparison,l: List[a]): Option[a]`

Add for other collections, including `Set` and `Map`. For `Nel` the method does not have to return an Option.

**Comment by mlutze** [comment]
Could this be done on a `Foldable` instead?

**Comment by magnus-madsen** [comment]
Yes, but maybe we want both. We will have both `List.map` and `Functor.map`.

**Comment by mlutze** [comment]
What do you see as the benefit of including both? My preference in general is to avoid having the programmer make arbitrary choices, since I think this would help maintain a standard style.

**Comment by magnus-madsen** [comment]
Scenario 1: No duplicates:

We move everything into instances. Thus there is no `Int32.minValue()` or `List.map`. You have to write `LowerBound.minValue(): Int32` or `Functor.map` (or use imports). This has the advantage that there is no duplication of names. On the other hand, not every single function will fit into a TC. Thus, when you go looking for a specific function you have to both consider *all* type classes* and e.g. the `Int32` or `List` namespace.

(The documentation for Haskell and Rust try to include instances in the docs for e.g. `Int32` or `List`.)

Scenario 2: We allow duplicates. (We do not duplicate implementations, but one will refer to the other).

If you are working with an `Int32` the recommendation is to use `Int32.minValue()`. If you are working with `List` the recommendation is to use `List.map`. Importantly: (a) everything related to `Int32` or `List` can be found in their respective namespaces. (b) You get better error messages because `List.map` is specific to `List`, so you don't get strange type errors. (c) You *do not need to know about TCs* to map over a list.

The last part is something I want to expand on. Maybe into a principle or blog post, but essentially it goes as follows:

In Flix:
- You can use `Option` without knowing about higher-order combinators.
- You can use `Option.map` without knowing about type classes.
- You can use `Functor` without knowing about category theory.

In other words, the language is "layered". You can start programming with just enums. Then you can add higher-order functions. Then you can add type classes. Then you can add type classes with higher-kinded types.

Thoughts?

**Comment by mlutze** [comment]
No rebuttal, I'm convinced :)
